### ISSUE TITLE
First call to AcquireTokenAsync fails running on Docker with .Net Core 2.2

### ISSUE THREAD

**Issue by rhallsie** [microsoftgraph/msgraph-sdk-dotnet#546]
Please provide the following (and please check them off the list with [x]) before submitting this issue:
- [x] Expected behavior. Please provide **links to the specific [Microsoft Graph documentation](https://developer.microsoft.com/en-us/graph/docs/concepts/overview)** you used to determine the expected behavior. 
- [x] Actual behavior. Provide error codes, stack information, and a [Fiddler](http://www.telerik.com/fiddler) capture of the request and response (please remove personally identifiable information before posting).
- [x] Steps to reproduce the behavior. Include your code, IDE versions, client library versions, and any other information that might be helpful to understand your scenario.

I have an implementation of IAuthenticationProvider that attempts to receive a token from https://graph.microsoft.com using client credentials. The code works fine running locally on Windows 10, but fails the first time it is hit when running in a docker container (subsequent requests succeed). The Windows environment is running .Net Core 2.2.401 and docker has version 2.2.7.

My implementation of IAuthenticationProvider has an AuthenticateRequestAsync method like this:
```
public async Task AuthenticateRequestAsync(HttpRequestMessage request)
{
    AuthenticationContext context = new 
    AuthenticationContext($"https://login.microsoftonline.com/{ADTENANT}", false);
    ClientCredential creds = new ClientCredential(APPID, SECRET);
    var authResult = await context.AcquireTokenAsync("https://graph.microsoft.com", creds);
    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
}
```

IAuthenticationProvider is configured in startup.cs like this:
```
services.AddTransient<Microsoft.Graph.IAuthenticationProvider, O365AuthenticationProvider>();
services.AddTransient<Microsoft.Graph.IGraphServiceClient, Microsoft.Graph.GraphServiceClient>();
```

### Expected behavior
The above call to  context.AcquireTokenAsync succeeds assuming the url and credentials are correct.

### Actual behavior
The call to  context.AcquireTokenAsync throws the following exception the first time it is called on docker:

```
Microsoft.Graph.ServiceException: Code: generalException
Message: An error occurred sending the request.
 ---> Microsoft.Graph.Auth.AuthenticationException: Code: generalException
Message: Unexpected exception occured while authenticating the request.
 ---> System.Net.Http.HttpRequestException: No such device or address ---> System.Net.Sockets.SocketException: No such device or address
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.AuthenticationHelper.SendWithAuthAsync(HttpRequestMessage request, Uri authUri, ICredentials credentials, Boolean preAuthenticate, Boolean isProxyAuth, Boolean doRequestAuth, HttpConnectionPool pool, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Identity.Client.Http.HttpManager.ExecuteAsync(Uri endpoint, IDictionary`2 headers, HttpContent body, HttpMethod method)
   at Microsoft.Identity.Client.Http.HttpManager.ExecuteWithRetryAsync(Uri endpoint, IDictionary`2 headers, HttpContent body, HttpMethod method, ICoreLogger logger, Boolean doNotThrow, Boolean retry)
   at Microsoft.Identity.Client.Http.HttpManager.SendGetAsync(Uri endpoint, IDictionary`2 headers, ICoreLogger logger)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.ExecuteRequestAsync[T](Uri endPoint, HttpMethod method, RequestContext requestContext)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.DiscoverAadInstanceAsync(Uri endPoint, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.NetworkMetadataProvider.SendInstanceDiscoveryRequestAsync(Uri authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.NetworkMetadataProvider.FetchAllDiscoveryMetadataAsync(Uri authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.NetworkMetadataProvider.GetMetadataAsync(Uri authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.InstanceDiscoveryManager.FetchNetworkMetadataOrFallbackAsync(RequestContext requestContext, Uri authorityUri)
   at Microsoft.Identity.Client.Instance.Discovery.InstanceDiscoveryManager.GetMetadataEntryAsync(String authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.UpdateAuthorityWithPreferredNetworkHostAsync()
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.ResolveAuthorityEndpointsAsync()
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenForClientParameters clientParameters, CancellationToken cancellationToken)
   at Microsoft.Graph.Auth.ClientCredentialProvider.AuthenticateRequestAsync(HttpRequestMessage httpRequestMessage)
   --- End of inner exception stack trace ---
   at Microsoft.Graph.Auth.ClientCredentialProvider.AuthenticateRequestAsync(HttpRequestMessage httpRequestMessage)
   at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.GraphServiceUsersCollectionRequest.GetAsync(CancellationToken cancellationToken)
```

### Steps to reproduce the behavior
1. Start a web API in docker with an endpoint that makes a call to GraphClient.Users.Request().Filter(filter).GetAsync().
2. The authentication provider should be configured as in the example above. 
3. Call the endpoint.

**Comment by darrelmiller** [comment]
I'm not sure what your error is, but we no longer recommend using ADAL for getting tokens for Microsoft Graph.  We use MSAL libraries now, and we have out of the box Authentication Providers for doing Client Credentials flow.  See this repo https://github.com/microsoftgraph/msgraph-sdk-dotnet-auth

Also, I wouldn't recommend using  AddTransient.  Both the Authentication Provider and GraphServiceClient are intended to be reused across the life of an application.

**Comment by rhallsie** [comment]
Thanks for the advice on lifespan and fetching tokens. I actually switched us over to ADAL yesterday but am experiencing the same issue. I've also tried singletons for both the graph client and authentication provider with no change. The issue appears to be happening at a lower level when it is trying to open an outbound socket for the first time. I don't think this is an issue with our containers because calls to other services work fine.
